### PR TITLE
pythonPackages.minio: init at 4.0.11

### PR DESCRIPTION
**Pull request by peterromfeldhk**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/minio/default.nix [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1zcz00i2f1zxfdy5gd0ynjv7wy9cdiqvr6xapjq516hjgwfqcl2v";
+  };
+
+  propagatedBuildInputs = [ urllib3 python-dateutil pytz ];
+
+  meta = with lib; {
+    description = "The Minio Python Client SDK provides simple APIs to access any Amazon S3 compatible object storage server";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/minio/default.nix]: Other meta attributes that are good to have are
- homepage
- maintainers



> https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

File: pkgs/development/python-modules/minio/default.nix [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{ stdenv, lib, buildPythonPackage, isPyPy, fetchPypi
+, urllib3, python-dateutil , pytz}:
+
+buildPythonPackage rec {
+  pname = "minio";
+  version = "4.0.11";
+
+  disabled = isPyPy;
+
+  src = fetchPypi {
```
Comment (worldofpeace) [inline pkgs/development/python-modules/minio/default.nix]: I see no tests in the source distribution.
Please fetch this from the [github repo](https://github.com/minio/minio-py) using `fetchFromGitHub`.

You'll also need to add the [testing requirements](https://github.com/minio/minio-py/blob/bd675fcca2b2e83f2f6b0a1de504a1b0abdb5ad0/setup.py#L49) to `checkInputs`.

Comment (peterromfeldhk) [reply pkgs/development/python-modules/minio/default.nix]: the tests run and passed without it?

**Comment by dotlambda** [comment]
@peterromfeldhk You are packaging multiple Python packages. Will this lead up to packaging some specific application? If so, please replace these PRs by a single one.

File: pkgs/development/python-modules/minio/default.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+buildPythonPackage rec {
+  pname = "minio";
+  version = "4.0.11";
+
+  disabled = isPyPy;
+
+  src = fetchFromGitHub {
+    owner = "minio";
+    repo = "${pname}-py";
+    rev = "${version}";
```
Comment (dotlambda) [inline pkgs/development/python-modules/minio/default.nix]:
```suggestion
    rev = version;
```

File: pkgs/development/python-modules/minio/default.nix [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    sha256 = "0rqma8iqwkmxrpmdnkrsr6425r9vgjnbw0j7rp1p13xb1alkjs6q";
+  };
+
+  propagatedBuildInputs = [ urllib3 python-dateutil pytz ];
+
+  meta = with lib; {
+    description = "Smple APIs to access any Amazon S3 compatible object storage server";
+    homepage = "https://github.com/minio/minio-py";
```
Comment (dotlambda) [inline pkgs/development/python-modules/minio/default.nix]:
```suggestion
    homepage = https://github.com/minio/minio-py;
```

**Review by dotlambda — CHANGES_REQUESTED** [review]
The commit message should equal this PR's title.

**Comment by peterromfeldhk** [comment]
> @peterromfeldhk You are packaging multiple Python packages. Will this lead up to packaging some specific application? If so, please replace these PRs by a single one.

oh ok.. will do.. yes the goal is to have all out project deps managed by nix to make easy nice containers :)

**Comment by peterromfeldhk** [comment]
@dotlambda not sure i understand you correctly... i am packaging up a lot of deps from a private project, so i can use nix in future. should i do one big MR?

**Comment by dotlambda** [comment]
> i am packaging up a lot of deps from a private project

In that case, there won't be an application I can test the dependencies with. Hence no need to do a single PR.

**Comment by flokli** [comment]
The minio python libraries are quite useful, and generic enough to have them packaged in nixpkgs IMHO.

I updated the minio test to use the minio python bindings to create a file and check for it afterwards:

```patch
From 74b37052c34d5d3265037fb74b3a23dd05fbbe4f Mon Sep 17 00:00:00 2001
From: Florian Klink <flokli@flokli.de>
Date: Tue, 5 Feb 2019 18:31:39 +0100
Subject: [PATCH] minio: test minio python library in minio test

---
 nixos/tests/minio.nix | 29 +++++++++++++++++++++++++----
 1 file changed, 25 insertions(+), 4 deletions(-)

diff --git a/nixos/tests/minio.nix b/nixos/tests/minio.nix
index 40a59954665..f1218b53771 100644
--- a/nixos/tests/minio.nix
+++ b/nixos/tests/minio.nix
@@ -1,4 +1,24 @@
-import ./make-test.nix ({ pkgs, ...} : {
+import ./make-test.nix ({ pkgs, ...} :
+let
+    accessKey = "BKIKJAA5BMMU2RHO6IBB";
+    secretKey = "V7f1CwQqAcwo80UEIJEjc5gVQUSSx5ohQ9GSrr12";
+    minioPythonScript = pkgs.writeScript "minio-test.py" ''
+      #! ${pkgs.python3.withPackages(ps: [ ps.minio ])}/bin/python
+      import io
+      import os
+      from minio import Minio
+      minioClient = Minio('localhost:9000',
+                    access_key='${accessKey}',
+                    secret_key='${secretKey}',
+                    secure=False)
+      sio = io.BytesIO()
+      sio.write(b'Test from Python')
+      sio.seek(0, os.SEEK_END)
+      sio_len = sio.tell()
+      sio.seek(0)
+      minioClient.put_object('test-bucket', 'test.txt', sio, sio_len, content_type='text/plain')
+    '';
+  in {
   name = "minio";
   meta = with pkgs.stdenv.lib.maintainers; {
     maintainers = [ bachp ];
@@ -8,8 +28,7 @@ import ./make-test.nix ({ pkgs, ...} : {
     machine = { pkgs, ... }: {
       services.minio = {
         enable = true;
-        accessKey = "BKIKJAA5BMMU2RHO6IBB";
-        secretKey = "V7f1CwQqAcwo80UEIJEjc5gVQUSSx5ohQ9GSrr12";
+        inherit accessKey secretKey;
       };
       environment.systemPackages = [ pkgs.minio-client ];
 
@@ -25,9 +44,11 @@ import ./make-test.nix ({ pkgs, ...} : {
       $machine->waitForOpenPort(9000);
 
       # Create a test bucket on the server
-      $machine->succeed("mc config host add minio http://localhost:9000 BKIKJAA5BMMU2RHO6IBB V7f1CwQqAcwo80UEIJEjc5gVQUSSx5ohQ9GSrr12 S3v4");
+      $machine->succeed("mc config host add minio http://localhost:9000 ${accessKey} ${secretKey} S3v4");
       $machine->succeed("mc mb minio/test-bucket");
+      $machine->succeed("${minioPythonScript}");
       $machine->succeed("mc ls minio") =~ /test-bucket/ or die;
+      $machine->succeed("mc cat minio/test-bucket/test.txt") =~ /Test from Python/ or die;
       $machine->shutdown;
 
     '';
-- 
2.19.2

```

@peterromfeldhk could you `git am` this patch and rebase?

**Comment by flokli** [comment]
@GrahamcOfBorg test minio

File: nixos/tests/minio.nix [hunk 5]
```diff
@@ -1,4 +1,24 @@
-import ./make-test.nix ({ pkgs, ...} : {
+import ./make-test.nix ({ pkgs, ...} :
+let
+    accessKey = "BKIKJAA5BMMU2RHO6IBB";
+    secretKey = "V7f1CwQqAcwo80UEIJEjc5gVQUSSx5ohQ9GSrr12";
+    minioPythonScript = pkgs.writeScript "minio-test.py" ''
+      #! ${pkgs.python3.withPackages(ps: [ ps.minio ])}/bin/python
```
Comment (dotlambda) [inline nixos/tests/minio.nix]:
```suggestion
      #! ${(pkgs.python3.withPackages (ps: [ ps.minio ])}).interpreter}
```

File: pkgs/development/python-modules/minio/default.nix [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    repo = "${pname}-py";
+    rev = version;
+    sha256 = "0rqma8iqwkmxrpmdnkrsr6425r9vgjnbw0j7rp1p13xb1alkjs6q";
+  };
+
+  checkInputs = [ faker mock nose ];
+  propagatedBuildInputs = [ urllib3 python-dateutil pytz ];
+
+  meta = with lib; {
+    description = "Smple APIs to access any Amazon S3 compatible object storage server";
```
Comment (dotlambda) [inline pkgs/development/python-modules/minio/default.nix]:
```suggestion
    description = "Simple APIs to access any Amazon S3 compatible object storage server";
```

File: pkgs/development/python-modules/minio/default.nix [hunk 7]
```diff
@@ -0,0 +1,24 @@
+{ stdenv, lib, buildPythonPackage, isPyPy, fetchFromGitHub
+, urllib3, python-dateutil , pytz, faker, mock, nose }:
+
+buildPythonPackage rec {
+  pname = "minio";
+  version = "4.0.11";
+
+  src = fetchFromGitHub {
```
Comment (dotlambda) [inline pkgs/development/python-modules/minio/default.nix]: Please add a comment on why you're not using `fetchPypi`.

File: pkgs/development/python-modules/minio/default.nix [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  };
+
+  checkInputs = [ faker mock nose ];
+  propagatedBuildInputs = [ urllib3 python-dateutil pytz ];
+
+  meta = with lib; {
+    description = "Smple APIs to access any Amazon S3 compatible object storage server";
+    homepage = https://github.com/minio/minio-py;
+    maintainers = with maintainers; [ peterromfeldhk ];
+    license = with licenses; [ asl20 ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/minio/default.nix]:
```suggestion
    license = licenses.asl20;
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/minio/default.nix (hunk 7 above)
Comment (flokli) [reply pkgs/development/python-modules/minio/default.nix]: bumped to 4.0.13 and fetched from pypi.

File: pkgs/development/python-modules/minio/default.nix (hunk 8 above)
Comment (flokli) [reply pkgs/development/python-modules/minio/default.nix]: fixed.

File: pkgs/development/python-modules/minio/default.nix (hunk 6 above)
Comment (flokli) [reply pkgs/development/python-modules/minio/default.nix]: fixed.

**Comment by flokli** [comment]
@GrahamcOfBorg test minio

**Comment by flokli** [comment]
@GrahamcOfBorg build nixosTests.minio

**Comment by dotlambda** [comment]
> The commit message should equal this PR's title.

This was not fixed.

**Comment by flokli** [comment]
My bad, sorry :-/

**Pull request merged**